### PR TITLE
style: Check transitions per longhand to know which transitions to keep.

### PR DESCRIPTION
**Pull request by emilio**
This fixes bug https://bugzilla.mozilla.org/show_bug.cgi?id=1406111

This fixes the fishy TransitionProperty mapping which I complained about in my
previous refactor.

Turns out that those properties could only be longhands, and thus the expansion
we did before that (and which I removed) was correct.

This fixes the bug by moving back to the previous correct behavior but using the
correct types.

The optimization to avoid creating a HashSet if we're transitioning all
properties or had no existing transition is removed since now we're creating a
LonghandIdSet, which is cheap, and that was only a performance optimization.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/18763)

### CODE REVIEW

**Comment by emilio** [comment]
r? @birtles 

Not sure what a good way to test this would be though.

**Comment by emilio** [comment]
(or @hiikezoe)

**Comment by hiikezoe** [comment]
The both commits look good to me.
One thins I am concerned (though it's an underlying issue)  Gecko_ElementTransitions_EndValueAt might return [an empty value](https://hg.mozilla.org/mozilla-central/file/19b32a138d08/layout/style/nsTransitionManager.h#l69) . @birtles knows something about this since it's introduced in [bug 1265611](https://bugzilla.mozilla.org/show_bug.cgi?id=1265611) .

**Comment by birtles** [comment]
Yeah, like @hiikezoe mentioned, `Gecko_ElementTransitions_EndValueAt` can return a null value. This is not expected to happen under normal circumstances, but it has happened in the past (specifically when we failed to ignore disabled properties when generating transitions based on inherited changes).

The again the existing code was doing `debug_assert!` followed by `unwrap()` so in such a situation it would panic in a release build just like this updated code (which uses `expect()`) will.

It would be better to keep the `debug_assert!` and just skip the value if it's empty.

File: components/style/gecko/wrapper.rs [hunk 1]
```diff
@@ -1303,22 +1303,27 @@ impl<'le> TElement for GeckoElement<'le> {
...
-        let mut map = FnvHashMap::with_capacity_and_hasher(collection_length, Default::default());
+            unsafe { Gecko_ElementTransitions_Length(self.0) } as usize;
+        let mut map = FnvHashMap::with_capacity_and_hasher(
+            collection_length,
+            Default::default()
+        );
+
         for i in 0..collection_length {
-            let (property, raw_end_value) = unsafe {
-                (Gecko_ElementTransitions_PropertyAt(self.0, i as usize).into(),
```
Comment (birtles) [inline components/style/gecko/wrapper.rs]: I believe this is the last use of `Gecko_ElementTransitions_PropertyAt` so we should drop it from `ServoBindings.cpp` etc.

**Review by birtles — CHANGES_REQUESTED** [review]
This changeset needs a description of the change (saying what it is fixing, how, why it's ok to drop the 'all'/no transitions optimization etc.)

File: components/style/gecko/wrapper.rs (hunk 1 above)
Comment (emilio) [reply components/style/gecko/wrapper.rs]: Yes, but that can land separately, since otherwise I need to land coordinately which is a pain.

Comment (birtles) [reply components/style/gecko/wrapper.rs]: Absolutely. I mostly just want to see the description of what the changeset fixes, how, and why it's ok to drop the 'all'/no transitions optimization.

**Comment by emilio** [comment]
I added a description to the first commit, and to the PR description so that it shows up in mozilla-central's commit.

**Comment by birtles** [comment]
Excellent, thank you! Apologies if that was there before and I simply didn't notice it (GitHub's review interface often confuses me).

It's the weekend here already but I'll try to look this over later tonight.

**Review by birtles — APPROVED** [review]
I honestly don't follow everything here since I don't know what the previous refactor did, but I trust you do and @hiikezoe has checked this over too.

**Pull request merged**